### DISCUSSION TITLE
Texture Color Channel Swizzle Being Ignored

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#288] [Q&A]
Would there by any reason the newer **texture swizzle** feature would be ignored in D3D12 mode on one computer, but works perfectly on another? I'm using the _Enabled_ state for the swizzle feature when requesting the device, and don't seem to be getting any type of errors or messages. And on both PCs, `GetDeviceInfo()` gives `DeviceInfo.Features.TextureComponentSwizzle == 1` (enabled) after device setup.

My app is trying to swizzle greyscale textures into RRR1, but still getting output of R001 on my Alienware laptop. The PC that is ignoring the swizzle is an Alienware R17, Intel(R) Core(TM) i7-6820HK CPU @ 2.70GHz   2.71 GHz, 32GB memory, Windows 10, running on a NVidia GeForce GTX 980M. Here are some of Diligent's outputs:

Alienware:
Diligent Engine: Info: D3D12-capable adapter found: Intel(R) HD Graphics 530 (128 MB)
Diligent Engine: Info: Max device shader model: 6_2

Edit: I just noticed it is using the Intel HD description for my graphics card above. My laptop automatically switches to its NVidia GeForce GTX 980M when games are run. I'm assuming it is switching, but the fact that it shows the onboard Intel card here is concerning. Could this be related to the issue? I seem to get pretty decent frame-rate that is relative to the desktop.

Dev PC:
Diligent Engine: Info: D3D12-capable adapter found: NVIDIA GeForce RTX 4090 (24156 MB)
Diligent Engine: Info: Max device shader model: 6_5

Any ideas what might be causing this, or anything I can do to investigate further, or narrow down the cause?

Here is the code that activates the view for the swizzle. I think most of it should read pretty easily, but let me know if anything is unclear:
```
if( GpuTex != null )
{
	// try to access the texture view
	GpuShaderView = GpuTex->GetDefaultView( Diligent::TEXTURE_VIEW_TYPE::TEXTURE_VIEW_SHADER_RESOURCE );
	
	// check for custom component swizzle
	if( GpuShaderView != null && textraits(texTrait::SWIZZLE) )
	{
		// verify situation
		_VE( texstats.Swizzle != TEX_SWIZZLE_DEFAULT, "swizzle at default?" );

		// prepare our shader view description as a copy of the default shader view
		Diligent::TextureViewDesc view_desc( GpuShaderView->GetDesc() );

		// set swizzles for texture view
		view_desc.Swizzle.R = TEXSWIZZLEMAP[ texstats.Swizzle[texSwizChan::R] ];
		view_desc.Swizzle.G = TEXSWIZZLEMAP[ texstats.Swizzle[texSwizChan::G] ];
		view_desc.Swizzle.B = TEXSWIZZLEMAP[ texstats.Swizzle[texSwizChan::B] ];
		view_desc.Swizzle.A = TEXSWIZZLEMAP[ texstats.Swizzle[texSwizChan::A] ];

		// create our shader view using this custom setup
		GpuShaderView = null;
		GpuTex->CreateView( view_desc, &GpuShaderView );
		_VE( GpuShaderView != null, "unable to create custom view, even though the default view was valid" );
	}
}
```
I appreciate any advice or information you might have!

**Comment by TheMostDiligent** [comment]
Can you try to run the texture swizzle unit test (use the `-DDILIGENT_BUILD_TESTS=ON` to build the DiligentCoreAPITest project and then run it with command arguments `--mode=d3d12 --gtest_filter=TextureSwizzleTest*`

**Comment by RobertMtx** [comment]
I haven't messed with your unit test system before, so I'm not sure I did this right, but I think I found and executed the right exe with the correct args. Let me know if not. If so, it looks like it passed on the laptop:
```

C:\Temp\DiligentCoreAPITest\Debug>DiligentCoreAPITest.exe --mode=d3d12 --gtest_filter=TextureSwizzleTest*
==================== Running tests in Direct3D12 mode ====================

Diligent Engine: Info: Found 3 compatible adapters
Diligent Engine: Info: Adapter 0: 'Intel(R) HD Graphics 530' (HW), Local/Host-Visible/Unified Memory: 128 MB / 16335 MB / 0 MB; 65 display modes
Diligent Engine: Info: Adapter 1: 'NVIDIA GeForce GTX 980M' (HW), Local/Host-Visible/Unified Memory: 8144 MB / 16335 MB / 0 MB
Diligent Engine: Info: Adapter 2: 'Microsoft Basic Render Driver' (SW), Local/Host-Visible/Unified Memory: 0 MB / 16335 MB / 0 MB
Diligent Engine: Info: User-defined allocator is not provided. Using default allocator.
Diligent Engine: Info: D3D12-capable adapter found: Intel(R) HD Graphics 530 (128 MB)
Diligent Engine: Info: User-defined allocator is not provided. Using default allocator.
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0xffffb80202840000
Diligent Engine: Info: Max device shader model: 6_2
Diligent Engine: Info: Adapter description: Intel(R) HD Graphics 530. Vendor: Intel. Local memory: 128 MB. Host-visible memory: 16335 MB. Unified memory: 0 MB.
Diligent Engine: Info: Selected shader compiler: Default
Note: Google Test filter = TextureSwizzleTest*
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from TextureSwizzleTest
[ RUN      ] TextureSwizzleTest.Red
[       OK ] TextureSwizzleTest.Red (48 ms)
[ RUN      ] TextureSwizzleTest.Green
[       OK ] TextureSwizzleTest.Green (46 ms)
[ RUN      ] TextureSwizzleTest.Blue
[       OK ] TextureSwizzleTest.Blue (48 ms)
[ RUN      ] TextureSwizzleTest.Alpha
[       OK ] TextureSwizzleTest.Alpha (49 ms)
[ RUN      ] TextureSwizzleTest.One
[       OK ] TextureSwizzleTest.One (46 ms)
[ RUN      ] TextureSwizzleTest.Zero
[       OK ] TextureSwizzleTest.Zero (43 ms)
[----------] 6 tests from TextureSwizzleTest (296 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (331 ms total)
[  PASSED  ] 6 tests.



Diligent Engine: Info: Device context stats
  Command counters
    SetPipelineState          12
    CommitShaderResources     12
    SetVertexBuffers          0
    SetIndexBuffer            0
    SetRenderTargets          12
    SetBlendFactors           0
    SetStencilRef             0
    SetViewports              12
    SetScissorRects           0
    ClearRenderTarget         12
    ClearDepthStencil         0
    Draw                      12
    DrawIndexed               0
    DrawIndirect              0
    DrawIndexedIndirect       0
    DispatchCompute           0
    DispatchComputeIndirect   0
    DispatchTile              0
    DrawMesh                  0
    DrawMeshIndirect          0
    MultiDraw                 0
    MultiDrawIndexed          0
    BuildBLAS                 0
    BuildTLAS                 0
    CopyBLAS                  0
    CopyTLAS                  0
    WriteBLASCompactedSize    0
    WriteTLASCompactedSize    0
    TraceRays                 0
    TraceRaysIndirect         0
    UpdateSBT                 0
    UpdateBuffer              0
    CopyBuffer                0
    MapBuffer                 12
    UpdateTexture             0
    CopyTexture               12
    MapTextureSubresource     12
    BeginQuery                0
    GenerateMips              0
    ResolveTextureSubresource 0
    BindSparseResourceMemory  0
  Primitives
    TRIANGLE_LIST             24
    TRIANGLE_STRIP            0
    POINT_LIST                0
    LINE_LIST                 0
    LINE_STRIP                0
    1_CONTROL_POINT_PATCHLIST 0
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 12/16
Diligent Engine: Info: Dynamic heap of immediate context usage stats:
                       Peak used/aligned/allocated size: 3.00 KB / 3.00 KB / 1.00 MB (1 page). Peak efficiency (used/aligned): 100.0%. Peak utilization (used/allocated): 0.3%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of immediate context usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of immediate context usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of immediate context usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of deferred context #4 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of deferred context #4 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of deferred context #4 usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of deferred context #5 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of deferred context #5 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of deferred context #5 usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of deferred context #6 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of deferred context #6 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of deferred context #6 usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of deferred context #7 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of deferred context #7 usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of deferred context #7 usage stats:
                       Peak used/aligned/allocated size: 0 Bytes / 0 Bytes / 0 Bytes (0 pages). Peak efficiency (used/aligned): 0.0%. Peak utilization (used/allocated): 0.0%
Diligent Engine: Info: Dynamic memory manager usage stats:
                       Total allocated memory: 1.00 MB
Diligent Engine: Info: D3D12 query manager peak usage:
QUERY_TYPE_OCCLUSION          :    0/ 128
QUERY_TYPE_BINARY_OCCLUSION   :    0/ 128
QUERY_TYPE_TIMESTAMP          :    0/ 512
QUERY_TYPE_PIPELINE_STATISTICS:    0/ 128
QUERY_TYPE_DURATION           :    0/ 512
Diligent Engine: Info: D3D12 query manager peak usage:
QUERY_TYPE_TIMESTAMP          :    0/ 512
QUERY_TYPE_DURATION           :    0/ 512
Diligent Engine: Info: D3D12 query manager peak usage:
Diligent Engine: Info: D3D12 query manager peak usage:
QUERY_TYPE_OCCLUSION          :    0/ 128
QUERY_TYPE_BINARY_OCCLUSION   :    0/ 128
QUERY_TYPE_TIMESTAMP          :    0/ 512
QUERY_TYPE_PIPELINE_STATISTICS:    0/ 128
QUERY_TYPE_DURATION           :    0/ 512
Diligent Engine: Info: Command list manager: created 1 allocators
Diligent Engine: Info: Command list manager: created 1 allocators
Diligent Engine: Info: Command list manager: created 3 allocators
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_SAMPLER     GPU heap max allocated size (static|dynamic): 0/1024 (0.00%) | 0/1024 (0.00%).
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV GPU heap max allocated size (static|dynamic): 0/16384 (0.00%) | 16/8192 (0.20%).
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_DSV         CPU heap allocated pool count: 1. Max descriptors: 1/16 (6.25%).
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_RTV         CPU heap allocated pool count: 1. Max descriptors: 9/16 (56.25%).
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_SAMPLER     CPU heap allocated pool count: 1. Max descriptors: 0/32 (0.00%).
Diligent Engine: Info: D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV CPU heap allocated pool count: 1. Max descriptors: 26/64 (40.62%).

```

**Reply by TheMostDiligent** [reply]
That looks correct.
Is this the same GPU where your swizzle fails?

**Reply by RobertMtx** [reply]
Yes. I'm looking into other reasons it might be failing. There are a few other variables that may be related, such as loading the project over a network, and the fact that I noticed something changed after updating NVidia drivers. It was happening with all greyscale textures. But after updating the drivers, only 2D images are now behaving this way. Its entirely possible I made some other change apart from the drivers that I'm not noticing, but still trying to narrow it down.

**Reply by TheMostDiligent** [reply]
TBH, I don't know what to suggest here. This feature does not require any synchronization or anything tricky.
Unit tests suggest GPU handles it properly unless there is some really tricky combination of swizzles.
It may be something unrelated to swizzle.

**Reply by TheMostDiligent** [reply]
FYI, besides run-time swizzle in the GPU, texture loader can perform [component swizzling](https://github.com/DiligentGraphics/DiligentTools/blob/da116e30adff75ccdb33443d604ca6d153ee1589/TextureLoader/interface/TextureLoader.h#L111) when loading textures.

**Reply by RobertMtx** [reply]
I have control over how the textures are saved, but mainly wanted to use single channel textures to save memory space when they are single-channel greyscale. Related to the problem, I'm still investigating. I'm hoping its a mistake on my part that can just be tweaked. If it works on one PC, but not another, it may be related to shader features, but no clue which ones. I'm requesting DX feature level 12, and was under the impression that largely means specific features are present, but its possible I'm doing something outside of those features without realizing it. Will post the reason it was happening once I figure it out, if I can. I really appreciate your help.

**Reply by TheMostDiligent** [reply]
I unfortunately don't have any suggestions here...

**Reply by RobertMtx** [reply]
I was wrong about it being exclusively 2D rendering affected. 3D is being affected the same way, but I was only reading the red channel of 3D textures in the shader and treating that value as a scaler, which made it seem as if 3D was correctly swizzled. So its actually all textures that are affected.

Here are some screenshots showing what the visuals look like in the map editor when the swizzle view is completely disabled (no secondary view is created) and then enabled normally. Both situations are "red" for the laptop, but the results look different. So it's not that swizzle isn't happening, but that it's happening incorrectly, or having some other affect. It's most noticeable in the bitmap fonts, which are written as single channel red textures, then swizzled to 111R. When swizzle is disabled, those fonts show up as expected on the laptop. But when its enabled, they show up as if their pixel value is always 1001 (opaque red).

So yeah, I have no idea what's going on. But I'm pretty sure it's out of my reach to fix. Let me know if you have any suggestions for tests, or anything that might narrow down what's happening. If you are curious to see any code or shaders, you are welcome to see it. I will probably keep investigating this until I run out of variables to verify.

No swizzle view, dev PC
![no_swizzle_view_desktop](https://github.com/DiligentGraphics/DiligentEngine/assets/43590381/94a3aeee-bd8b-464c-a89d-fa956def185d)

No swizzle view, laptop
![no_swizzle_view_laptop](https://github.com/DiligentGraphics/DiligentEngine/assets/43590381/c024bf40-c620-4055-a3b2-63ce668f7399)

Normal swizzle view, dev PC
![swizzle_view_desktop](https://github.com/DiligentGraphics/DiligentEngine/assets/43590381/0c27030e-4e6d-4d1d-89cc-e5545d59fbaf)

Normal swizzle view, laptop
![swizzle_view_laptop](https://github.com/DiligentGraphics/DiligentEngine/assets/43590381/1ef50635-4a61-4349-99d2-68fd97a1d8fd)

**Reply by TheMostDiligent** [reply]
I suggest experimenting with one of Diligent Tutorials - e.g. Tutorial03. Try using the same texture swizzle as you use in your app and see if that works.

**Reply by RobertMtx** [reply]
When I read this, I thought "That's a lot of work", but then did it in less than 5 minutes. I changed the DG logo to greyscale, created a view to swizzle the red channel into green, and it worked totally fine on both computers. Even though this seemed like an obvious thing to try, I appreciate you pushing me into it. Looks like I need to keep digging.

**Reply by RobertMtx** [reply]
I've run out of things to try, that won't require extensive work, such as rebuilding the sample framework to do things exactly like my engine does. I did everything I could imagine to make my app load and draw textures exactly like the sample framework, but to no effect. The only thing that makes swizzle work correctly (in my app, on the laptop) is forcing the format of the texture resource to be 4-channel RGBA. When the texture uses `TEXTURE_FORMAT::TEX_FORMAT_R8_UNORM` (red), it is always some level of red, and swizzle doesn't seem capable of changing any of the GBA channels, but *can* change the red channel. I've tried loading as `.png` and `.dds`, but both fail as single-channel textures.

The sample framework does work with `TEXTURE_FORMAT::TEX_FORMAT_R8_UNORM` and swizzle on the laptop. Since I've already tried forcing my app to load and draw exactly the same way as the samples, I'm guessing this is being caused by some larger setting or state somewhere, like the primary graphics settings. Anyway, I've decided to take a break from this nightmare to make some actual progress, so figured I would mention that. Sometimes my brain will find solutions once I stop caring. Thanks again for your help.

**Reply by TheMostDiligent** [reply]
Did you try to capture your frame in RenderDoc or any other Graphics debugger tool (e.g. Visual Studio Graphics Debugger) to see what is happening?
